### PR TITLE
Prevent stray output from reaching stdout

### PR DESCRIPTION
**Pull request by phiSgr**
prevents prints that happen after (user's) `main` `return`s from polluting stdout.

Context: https://stackoverflow.com/questions/75714151/kotlin-coroutines-order-of-printing-logs/75760394#75760394

### CODE REVIEW

**Comment by phiSgr** [comment]
On second thought, not switching System.out back may cause stray prints become data races as the main thread reads from `ByteArrayOutputStream` and the thread the prints writes to it.

But if that is the reason for calling `setOut`, `System.err` should also be switched back.

---

Or if the writes synchronize over the BAOS as I suggested in #619, the data race can be prevented by

```kt
synchronized(outputStream) {
    outputObj.text = outputStream.toString()
        .replace("</errStream><errStream>".toRegex(), "")
        .replace("</outStream><outStream>".toRegex(), "")
}
```

**Comment by AlexanderPrendota** [comment]
Hi, is it possible to add a test snippet?

File: src/test/kotlin/com/compiler/server/CoroutinesRunnerTest.kt [hunk 1]
```diff
@@ -57,10 +57,11 @@ class CoroutinesRunnerTest : BaseExecutorTest() {
 
   @Test
   fun `base coroutines test 7`() {
-    run(
+    val result = run(
       code = "import kotlinx.coroutines.*\n\nfun main() = runBlocking {\nGlobalScope.launch {\n    repeat(1000) { i ->\n        println(\"I'm sleeping \$i ...\")\n        delay(500L)\n    }\n}\ndelay(1300L) // just quit after delay    \n}",
-      contains = "I'm sleeping 0 ...\nI'm sleeping 1 ...\nI'm sleeping 2 ...\n"
+      contains = ""
     )
+    Assertions.assertEquals("<outStream>I'm sleeping 0 ...\nI'm sleeping 1 ...\nI'm sleeping 2 ...\n</outStream>", result.text)
```
Comment (phiSgr) [inline src/test/kotlin/com/compiler/server/CoroutinesRunnerTest.kt:64]: If you go before the fix with `git checkout 5e25435`, this test will intermittently fail.

File: src/main/kotlin/com/compiler/server/model/ProgramOutput.kt [hunk 2]
```diff
@@ -26,14 +26,7 @@ data class ProgramOutput(
       standardOutput.isBlank() -> ExecutionResult()
       else -> {
         try {
-          // coroutines can produced incorrect output. see example in `base coroutines test 7`
-          if (standardOutput.startsWith("{")) outputMapper.readValue(standardOutput, ExecutionResult::class.java)
-          else {
-            val result = outputMapper.readValue("{" + standardOutput.substringAfter("{"), ExecutionResult::class.java)
```
Comment (phiSgr) [inline src/main/kotlin/com/compiler/server/model/ProgramOutput.kt:32]: I believe we won't need this workaround, but I wouldn't object to keeping it.

File: executors/src/main/kotlin/JUnitExecutors.kt [hunk 3]
```diff
@@ -29,13 +29,12 @@ class JUnitExecutors {
           }
           groupedTestResults[testRunInfo.className]?.add(testRunInfo)
```
Comment (phiSgr) [inline executors/src/main/kotlin/JUnitExecutors.kt:30]: Replace with `.groupBy`?

File: executors/src/main/kotlin/JUnitExecutors.kt [hunk 4]
```diff
@@ -29,13 +29,12 @@ class JUnitExecutors {
...
         }
-        print(mapper.writeValueAsString(groupedTestResults))
+        standardOutput.print(mapper.writeValueAsString(groupedTestResults))
       }
       catch (e: Exception) {
-        System.setOut(standardOutput)
-        print("[\"")
+        standardOutput.print("[\"")
         e.printStackTrace()
```
Comment (phiSgr) [inline executors/src/main/kotlin/JUnitExecutors.kt:36]: It seems to me that it's writing to `PrintStream(errorStream)`. Then the BAOS is discarded. Should it be sent to stderr?

**Comment by phiSgr** [comment]
> possible to add a test

Is the modified test sufficient?
https://github.com/JetBrains/kotlin-compiler-server/pull/618#discussion_r1148465908

**Comment by phiSgr** [comment]
Following up on this.

I wasn't very clear in my previous comment. I have changed the test which now guards against the bug. I don't think a new additional test is needed.

The test that I modified, `base coroutines test 7` used to only test for substring match.
Now it is testing for full equality.

**Comment by phiSgr** [comment]
This bug has confused another person.

https://stackoverflow.com/questions/77619172/why-does-the-kotlin-coroutine-output-a-b-even-with-a-delay

**Comment by phiSgr** [comment]
I found another SO question caused by this bug.
Imagine how many people are confused by this but didn't ask a question.

https://stackoverflow.com/questions/71918773/why-does-coroutine-in-background-execute-before-the-main-thread-in-this-kotlin-c

**Comment by nikpachoo** [comment]
@phiSgr, thank you for bringing this up! 
We are reviewing the PR, and I hope we can merge it this week.

File: executors/src/main/kotlin/JUnitExecutors.kt (hunk 3 above)
Comment (nikpachoo) [reply executors/src/main/kotlin/JUnitExecutors.kt:30]: Thanks for the suggestion, replaced.

File: executors/src/main/kotlin/JUnitExecutors.kt (hunk 4 above)
Comment (nikpachoo) [reply executors/src/main/kotlin/JUnitExecutors.kt:36]: Well, hard to say. The code in the `catch` block seems nonfunctional, regardless of the changes. I tried to reproduce a case when the execution flow reached this part of the code and got a serialization error.

**Review by nikpachoo — APPROVED** [review]

**Comment by nikpachoo** [comment]
No objections from my side.
@phiSgr, thank you for the contribution!

**Comment by phiSgr** [comment]
It seems to me that it's now deployed to play.kotlinlang.org only in version 2.0.0-RC2. Will it be back-ported to older versions as well?

**Comment by nikpachoo** [comment]
@phiSgr, 
1.9.24 version released today contains the fix ✅ 
There are no plans to back-port it to all the previous versions; their usage is relatively low.
Thank you for your participation ❤️

**Pull request merged**